File: bindings/fortran/src/safe_string.f90
! SPDX-License-Identifier: PMPL-1.0-or-later
! Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
!
! Proven SafeString - XSS prevention for Fortran
!
! Thin wrapper around libproven's verified SafeString module.
! All escaping logic is executed in the Idris 2 core via the Zig FFI
! bridge; this module only marshals Fortran strings to/from C pointers.
!
! Link with: -lproven

module safe_string
    use, intrinsic :: iso_c_binding
    use proven_ffi
    implicit none
    private

    public :: escape_html, escape_sql, escape_js, is_valid_utf8

contains

    ! =========================================================================
    ! Helper: convert a Fortran string into a C pointer + length, call an
    !         FFI function that returns c_proven_string_result, and convert
    !         the result back to an allocatable Fortran string.
    ! =========================================================================

    !> Escape HTML special characters via proven_string_escape_html.
    !> Caller receives an allocatable Fortran string; libproven memory is freed
    !> before returning.
    function escape_html(input) result(output)
        character(len=*), intent(in) :: input
        character(len=:), allocatable :: output
        character(len=len_trim(input), kind=c_char), target :: c_buf
        type(c_proven_string_result) :: c_res
        character(len=1, kind=c_char), pointer :: f_ptr(:)
        integer :: trimmed_len

        trimmed_len = len_trim(input)
        c_buf = input(1:trimmed_len)

        c_res = proven_string_escape_html(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK .and. c_associated(c_res%value)) then
            call c_f_pointer(c_res%value, f_ptr, [c_res%length])
            allocate(character(len=c_res%length) :: output)
            block
                integer :: i
                do i = 1, int(c_res%length)
                    output(i:i) = f_ptr(i)
                end do
            end block
            call proven_free_string(c_res%value)
        else
            output = input(1:trimmed_len)
        end if
    end function escape_html

    !> Escape SQL single-quote characters via proven_string_escape_sql.
    function escape_sql(input) result(output)
        character(len=*), intent(in) :: input
        character(len=:), allocatable :: output
        character(len=len_trim(input), kind=c_char), target :: c_buf
        type(c_proven_string_result) :: c_res
        character(len=1, kind=c_char), pointer :: f_ptr(:)
        integer :: trimmed_len

        trimmed_len = len_trim(input)
        c_buf = input(1:trimmed_len)

        c_res = proven_string_escape_sql(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK .and. c_associated(c_res%value)) then
            call c_f_pointer(c_res%value, f_ptr, [c_res%length])
            allocate(character(len=c_res%length) :: output)
            block
                integer :: i
                do i = 1, int(c_res%length)
                    output(i:i) = f_ptr(i)
                end do
            end block
            call proven_free_string(c_res%value)
        else
            output = input(1:trimmed_len)
        end if
    end function escape_sql

    !> Escape JavaScript string literal characters via proven_string_escape_js.
    function escape_js(input) result(output)
        character(len=*), intent(in) :: input
        character(len=:), allocatable :: output
        character(len=len_trim(input), kind=c_char), target :: c_buf
        type(c_proven_string_result) :: c_res
        character(len=1, kind=c_char), pointer :: f_ptr(:)
        integer :: trimmed_len

        trimmed_len = len_trim(input)
        c_buf = input(1:trimmed_len)

        c_res = proven_string_escape_js(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK .and. c_associated(c_res%value)) then
            call c_f_pointer(c_res%value, f_ptr, [c_res%length])
            allocate(character(len=c_res%length) :: output)
            block
                integer :: i
                do i = 1, int(c_res%length)
                    output(i:i) = f_ptr(i)
                end do
            end block
            call proven_free_string(c_res%value)
        else
            output = input(1:trimmed_len)
        end if
    end function escape_js

    !> Check if a byte sequence is valid UTF-8 via proven_string_is_valid_utf8.
    function is_valid_utf8(input) result(valid)
        character(len=*), intent(in) :: input
        logical :: valid
        character(len=len_trim(input), kind=c_char), target :: c_buf
        type(c_proven_bool_result) :: c_res
        integer :: trimmed_len

        trimmed_len = len_trim(input)
        c_buf = input(1:trimmed_len)

        c_res = proven_string_is_valid_utf8(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK) then
            valid = logical(c_res%value)
        else
            valid = .false.
        end if
    end function is_valid_utf8

end module safe_string

File: bindings/pascal/src/SafeString.pas
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

// SafeString - Safe string operations wrapper for libproven.
//
// Provides UTF-8 validation, SQL/HTML/JS escaping via the formally verified
// Idris 2 core. This unit ONLY wraps FFI calls; it does NOT reimplement
// any logic.
//
// Compatible with Free Pascal (FPC) and Delphi.

unit SafeString;

{$IFDEF FPC}
  {$MODE OBJFPC}{$H+}
{$ENDIF}

interface

uses
  LibProven;

type
  // High-level result for safe string operations.
  // Check Success before accessing Value.
  TSafeStringResult = record
    Success: Boolean;
    Value: AnsiString;
    ErrorCode: Int32;
  end;

  // High-level result for boolean validation operations.
  TSafeBoolResult = record
    Success: Boolean;
    Value: Boolean;
    ErrorCode: Int32;
  end;

// Check if a string contains valid UTF-8 data.
// Calls proven_string_is_valid_utf8 via FFI.
function IsValidUTF8(const S: AnsiString): TSafeBoolResult;

// Escape a string for safe SQL embedding (single-quote escaping).
// Prefer parameterized queries over string escaping.
// Calls proven_string_escape_sql via FFI.
function EscapeSQL(const S: AnsiString): TSafeStringResult;

// Escape a string for safe HTML embedding (prevents XSS).
// Calls proven_string_escape_html via FFI.
function EscapeHTML(const S: AnsiString): TSafeStringResult;

// Escape a string for safe JavaScript string literal embedding.
// Calls proven_string_escape_js via FFI.
function EscapeJS(const S: AnsiString): TSafeStringResult;

implementation

// Convert raw FFI string result to high-level Pascal result.
// Frees the C-allocated string after copying into a Pascal AnsiString.
function ToSafeStringResult(const R: TProvenStringResult): TSafeStringResult;
begin
  Result.Success := (R.Status = PROVEN_OK);
  if Result.Success then
  begin
    SetString(Result.Value, R.Value, R.Length);
    Result.ErrorCode := PROVEN_OK;
  end
  else
  begin
    Result.Value := '';
    Result.ErrorCode := R.Status;
  end;
  // Always free the C-allocated string if it was allocated.
  if R.Value <> nil then
    proven_free_string(R.Value);
end;

// Convert raw FFI bool result to high-level Pascal result.
function ToSafeBoolResult(const R: TProvenBoolResult): TSafeBoolResult;
begin
  Result.Success := (R.Status = PROVEN_OK);
  if Result.Success then
  begin
    Result.Value := Boolean(R.Value);
    Result.ErrorCode := PROVEN_OK;
  end
  else
  begin
    Result.Value := False;
    Result.ErrorCode := R.Status;
  end;
end;

function IsValidUTF8(const S: AnsiString): TSafeBoolResult;
begin
  if Length(S) = 0 then
  begin
    // Empty strings are trivially valid UTF-8.
    Result.Success := True;
    Result.Value := True;
    Result.ErrorCode := PROVEN_OK;
    Exit;
  end;
  Result := ToSafeBoolResult(
    proven_string_is_valid_utf8(PByte(PAnsiChar(S)), NativeUInt(Length(S)))
  );
end;

function EscapeSQL(const S: AnsiString): TSafeStringResult;
begin
  if Length(S) = 0 then
  begin
    Result.Success := True;
    Result.Value := '';
    Result.ErrorCode := PROVEN_OK;
    Exit;
  end;
  Result := ToSafeStringResult(
    proven_string_escape_sql(PByte(PAnsiChar(S)), NativeUInt(Length(S)))
  );
end;

function EscapeHTML(const S: AnsiString): TSafeStringResult;
begin
  if Length(S) = 0 then
  begin
    Result.Success := True;
    Result.Value := '';
    Result.ErrorCode := PROVEN_OK;
    Exit;
  end;
  Result := ToSafeStringResult(
    proven_string_escape_html(PByte(PAnsiChar(S)), NativeUInt(Length(S)))
  );
end;

function EscapeJS(const S: AnsiString): TSafeStringResult;
begin
  if Length(S) = 0 then
  begin
    Result.Success := True;
    Result.Value := '';
    Result.ErrorCode := PROVEN_OK;
    Exit;
  end;
  Result := ToSafeStringResult(
    proven_string_escape_js(PByte(PAnsiChar(S)), NativeUInt(Length(S)))
  );
end;

end.
